controllers/cache/readiness.go

```diff
@@ -0,0 +1,264 @@
+		}
+
+		endpoint, port = GetExternalRedisHost(spec)
+
+		if spec.SecretName != "" {
+			pw, err = GetExternalRedisPassword(spec, redis.Namespace)
+		}
+
+		connect = &RedisConnect{
+			Endpoint:  strings.Join(endpoint[:], ","),
+			Port:      port,
+			Password:  pw,
+			GroupName: spec.GroupName,
+		}
+
+		redis.RedisConnect = connect
+		client = connect.NewRedisPool()
+	case "redis":
+		if len(spec.Hosts) != 1 {
+			return nil, errors.New(".redis.spec.hosts is invalid")
+		}
+		endpoint, port = GetExternalRedisHost(spec)
+
+		if spec.SecretName != "" {
+			pw, err = GetExternalRedisPassword(spec, redis.Namespace)
+		}
+
+		connect = &RedisConnect{
+			Endpoint:  fmt.Sprintf("%s:%s", endpoint, port),
+			Port:      port,
+			Password:  pw,
+			GroupName: spec.GroupName,
+		}
+		redis.RedisConnect = connect
+		client = connect.NewRedisClient()
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+// GetExternalRedisHost returns external redis host list and port
+func GetExternalRedisHost(spec *goharborv1.RedisSpec) ([]string, string) {
+	var (
+		endpoint []string
+		port     string
+	)
+	for _, host := range spec.Hosts {
+		sp := host.Host + ":" + host.Port
+		endpoint = append(endpoint, sp)
+		port = host.Port
+	}
+	return endpoint, port
+}
+
+// GetExternalRedisPassword returns external redis password
+func GetExternalRedisPassword(spec *goharborv1.RedisSpec, namespace string) (string, error) {
+	external := &RedisReconciler{
+		Name:      spec.SecretName,
+		Namespace: namespace,
+	}
+
+	pw, err := external.GetRedisPassword()
+	if err != nil {
+		return "", err
+	}
+
+	return pw, err
+}
+
+// GetInClusterRedisInfo returns inCluster redis sentinel pool client
+func (redis *RedisReconciler) GetInClusterRedisInfo() (*rediscli.Client, error) {
+	password, err := redis.GetRedisPassword()
+	if err != nil {
+		return nil, err
+	}
+	redis.Log.Info("Redis password.", "password", password)
+
+	_, sentinelPodList, err := redis.GetDeploymentPods()
+	if err != nil {
+		redis.Log.Error(err, "Fail to get deployment pods.")
+		return nil, err
+	}
+
+	_, redisPodList, err := redis.GetStatefulSetPods()
+	if err != nil {
+		redis.Log.Error(err, "Fail to get deployment pods.")
+		return nil, err
+	}
+
+	if len(sentinelPodList.Items) == 0 || len(redisPodList.Items) == 0 {
+		redis.Log.Info("pod list is empty，pls wait.")
+		return nil, nil
+	}
+
+	sentinelPodArray := sentinelPodList.Items
+
+	_, currentSentinelPods := redis.GetPodsStatus(sentinelPodArray)
+
+	if len(currentSentinelPods) == 0 {
+		return nil, errors.New("need to requeue")
+	}
+
+	endpoint := redis.GetSentinelServiceUrl(currentSentinelPods)
+
+	connect := &RedisConnect{
+		Endpoint:  endpoint,
+		Port:      RedisSentinelConnPort,
+		Password:  password,
+		GroupName: RedisSentinelConnGroup,
+	}
+
+	redis.RedisConnect = connect
+
+	client := connect.NewRedisPool()
+
+	return client, nil
+}
+
+func cacheNotReadyStatus(reason, message string) *lcm.CRStatus {
```

Comment:
Also left comments in  @wangcanfengxs's PR. Maybe we can have a better lcm.CRStatus that support 

`lcm.New().WithReason().WithMessage().WithProperties()`